### DISCUSSION TITLE
Logging conditionals

### DISCUSSION THREAD

**Discussion by ghost** [btraceio/btrace#545] [Q&A] [answered]
Hi,

Logging method entry & exit points is already quite useful.
But sometimes it's also useful to log the if/while and potential other constructs in the code of methods.

Is there any way to achieve this?

**Comment by jbachorik** [comment, marked as answer]
This would be quite difficult to achieve. There is no realistic way to identify a conditional or loop apart of the line number which might be matching also other code on the same line. 

What I can imagine is a special tracing mode where all conditionals (loops are in the end transformed to conditional jumps as well) are instrumented to capture the defined context (instance fields, local variables etc.) and store it in an efficient form. 

Unfortunately, this is not something BTrace could do OOTB - mostly because this mode of tracing will severely affect the application performance and is more like debugging than a light-weight tracing :(

**Reply by ghost** [reply]
Well, the whole point is enabling debugging through logs.
I've gotten quite used to logging all the decisions the code takes, so that i can easily understand how it executes at runtime; this drops diag cost dramatically.

While I care ab performance, the impact of such logging is minuscule for the regular use cases I manage; the bigger problem is the volume of logs tbh. Also, with hot (re)config, enabling this only when needed, this can be kept under control.

I think the bare minimum would be to just know a specific code path was taken (enter an *if* or a *catch*, like what jacoco is doing; logging the value of whatever was in the while/if, is less important and can be done by a human if needed.

**Reply by jbachorik** [reply]
I see your point. However, this would require a slightly different mode of work - the tracing should be fully automated without any user specified code (unlike other BTrace probes)
